text/0148-sampling-probability.md

```diff
@@ -0,0 +1,744 @@
+count; we can use this to combine multiple sampling schemes in a
+single stream.
+
+For example, consider collecting two samples from a stream of spans,
+the first sample from those spans with attribute `A=1` and the second
+sample from those spans with attribute `B=2`.  Any span that has both
+of these properties is eligible to be selected for both samples, in
+which case one event could have two non-zero adjusted counts (e.g.,
+`sampling.by_a.adjusted_count` and `sampling.by_b.adjusted_count`).
+
+## Proposed Tracing specification
+
+For the standard OpenTelemetry Span Sampler implementations to support
+a range of probability sampling schemes, this document recommends the
+use of a Span attribute named `sampling.sampler_name.adjusted_count`
+to encode an unbiased adjusted count computed by a Sampler.
+
+The value any attribute name prefixed with `sampling.` and suffixed
+with `.adjusted_count` under this proposal MUST be an unbiased
+estimate of the total population count represented by the individual
+event.
+
+### Suggested text
+
+After
+[Sampler](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler)
+and before [Builtin
+Samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-samplers),
+a new section will introduce the term "adjusted count" and relate it
+to inclusion probability.  For example:
+
+```markdown
+### Probability sampling
+
+Probability Samplers are Samplers that output statistically unbiased
+inclusion probability.  Inclusion probability in the context of
+tracing is the *effective* probability of the Sampler returning
+`RECORD_AND_SAMPLE` when invoked, which can be decided locally or 
+derived from the context when the [W3C Trace Context is-sampled
+flag](https://www.w3.org/TR/trace-context/#sampled-flag) is in use.
+
+#### Adjusted count span attribute
+
+The recommended way to convey inclusion probability *for events*
+sampled in OpenTelemetry is in the form of an **adjusted count**,
+which is the reciprocal (i.e., mathematical inverse function) of
+inclusion probability.
+
+The implied goal of probability sampling is to support estimating the
+count of spans in the population using the spans that were sampled.
+The adjusted count associated with a Span is the expected value of the
+number of identical Spans within the population that each Span
+represents.  Probability samplers SHOULD ensure the samples they
+compute are unbiased, which implies that the expected value of the sum
+of adjusted counts in the sample equals the true count of spans in the
+population.
+
+Attributes used to express the adjusted count in an unbiased
+probability sampling scheme SHOULD use a Span attribute name with
+prefix `sampling.` and with suffix `.adjusted_count` (e.g.,
+`sampling.sampler_name.adjusted_count`).  Adjusted count attributes
+MAY be integer or floating-point values.
+
+#### Inclusion probability tracestate value
```

Comment:
I think this should be optional

Reply:
Can you clarify @yurishkuro ? It seems essential to me to convey the probability in _some_ form when exporting   to consumers of span data (like the collector  and it's various exporters)? Otherwise how can those exporters know how to upscale any statistics or metrics?

Reply:
"essential to me" is the key part. This whole process (such as generating metrics on the fly from spans) may not be essential to some users, so the behavior should be opt-in.

Reply:
> "some". 

I think _a majority_ of users would find this useful. It's not _just_ on the fly metric generation. Including the sampling rate _somewhere_ in the exported payload means providing users a more complete picture of the sampling decision that occurred at the time in the service/application/etc. Not doing that means we're giving them an incomplete picture. Doing on the fly metric generation is just one way that can then be used down the line. 

Could you give an example of a user who would prefer the tradeoff off off-by-default incomplete/missing sampling information for the benefit of not having to include/compute the extra key/value pair in their span data? I'm genuinely not trying to be snarky here I just am not sure I understand the "user story", so to speak. The use case I can think of (web-scale/perf costs/reduce payload size/etc) feels better served by `on-by-default, optional opt-out`.

Obviously I don't have the depth of knowledge you do here, but I think of sampling as a 1st class citizen in _tracing_ and what separates it from just structured logging, and so we should aim for as rich a sampling picture as is reasonable.

Anyway, hope that helps clarify what "essential to me" means.

Reply:
Jaeger, Zipkin, OpenCensus SDKs do not propagate sampling weight. W3C Trace Context does not include the weight as a required part of the context. I think this provides enough evidence that this feature is not a requirement for a functioning tracing system.

I am completely on board with capturing the weight as a span attribute - all Jaeger SDKs do that, and any post-collection processing that hopes to build any aggregate data from traces would need that. But in-band propagation is not required for most use cases.

Reply:
yup, per usual i'm an idiot and misread the previous section. plain ole span attribute works just fine for me, in-band (traceState) far less important and i don't have strong opinions here. txs for the patience.

Reply:
Having laid out the argument for propagating head inclusion probability, I can agree we should all be on the fence about it.

If the Parent sampler will be used and head inclusion probability is not propagated when the is-sampled bit is set, we will have a difficult time computing metrics from spans. However, the probability is only needed when is-sampled is true, so by turning on sampling the cost of propagating the sampling rate is automatically lowered.

Only the Dapper "inflationary" approach requires propagating the inclusion probability for negative sampling decisions, and that approach is not one I'm trying to defend. Unfortunately, we are left with few options that meet the three requirements stated in this document, which are:

1. Reduces Tracer overhead
2. Produces complete traces
3. Spans are countable.

OpenCensus was trying to switch to the use of TraceIDRatio sampling by default, which satisfies 1 and 3 easily. OpenTelemetry has a problem with 2, but I'm not sure it has to be solved. Completeness is less important than spans-to-metrics support, from our perspective as a vendor.

Reply:
I replaced the MUST with MAY to signify this is optional support.